// remix/server.js
/**
 * @remix-run/server-runtime v1.0.6
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var serverRuntime = require('@remix-run/server-runtime');



Object.defineProperty(exports, 'createCookie', {
  enumerable: true,
  get: function () { return serverRuntime.createCookie; }
});
Object.defineProperty(exports, 'createCookieSessionStorage', {
  enumerable: true,
  get: function () { return serverRuntime.createCookieSessionStorage; }
});
Object.defineProperty(exports, 'createMemorySessionStorage', {
  enumerable: true,
  get: function () { return serverRuntime.createMemorySessionStorage; }
});
Object.defineProperty(exports, 'createSession', {
  enumerable: true,
  get: function () { return serverRuntime.createSession; }
});
Object.defineProperty(exports, 'createSessionStorage', {
  enumerable: true,
  get: function () { return serverRuntime.createSessionStorage; }
});
Object.defineProperty(exports, 'isCookie', {
  enumerable: true,
  get: function () { return serverRuntime.isCookie; }
});
Object.defineProperty(exports, 'isSession', {
  enumerable: true,
  get: function () { return serverRuntime.isSession; }
});
Object.defineProperty(exports, 'json', {
  enumerable: true,
  get: function () { return serverRuntime.json; }
});
Object.defineProperty(exports, 'redirect', {
  enumerable: true,
  get: function () { return serverRuntime.redirect; }
});
